compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h

```diff
@@ -64,38 +65,51 @@ void eraseTranslationInfo(mlir::FunctionOpInterface funcOp);
 // operations.
 //===----------------------------------------------------------------------===//
 
+static const char kConfigAttrName[] = "lowering_config";
+
+/// Returns the lowering configuration set for an operation. Returns `nullptr`
+/// if no value is set.  It expects that the attribute is stored using the
+/// identifier `lowering_config`.
+template <typename ConfigTy = IREE::Codegen::LoweringConfigAttrInterface>
+ConfigTy getLoweringConfig(Operation *op) {
+  return op->getAttrOfType<ConfigTy>(kConfigAttrName);
+}
+
 /// Returns the op that carries the `lowering_config` attribute; returns nullptr
 /// if none is carrying the attribute.
 ///
 /// This scans ops in top-down order and the first one carrying the attribute
 /// will be returned.
+template <typename ConfigTy = IREE::Codegen::LoweringConfigAttrInterface>
 FailureOr<Operation *>
-getLoweringConfigCarryingOp(ArrayRef<Operation *> computeOps);
-
-/// Returns the lowering configuration set for an operation. Returns `nullptr`
-/// if no value is set.  It expects that the attribute is stored using the
-/// identifier `lowering_config`.
-IREE::Codegen::LoweringConfigAttr getLoweringConfig(Operation *op);
+getLoweringConfigCarryingOp(ArrayRef<Operation *> computeOps) {
+  for (Operation *op : computeOps) {
+    if (getLoweringConfig<ConfigTy>(op))
+      return op;
+  }
+  return failure();
+}
 
 /// Returns the lowering configuration from the list of operations; returns
 /// nullptr if unable to find.
 ///
 /// This scans ops in top-down order and the first one carrying the attribute
 /// will be returned.
-FailureOr<IREE::Codegen::LoweringConfigAttr>
-getLoweringConfig(ArrayRef<Operation *> computeOps);
+template <typename ConfigTy = IREE::Codegen::LoweringConfigAttrInterface>
+FailureOr<ConfigTy> getLoweringConfig(ArrayRef<Operation *> computeOps) {
```

Comment:
I think this might cause some non-trivial name conflicts. `Operation *` can be implicitly converted to `ArrayRef<Operation *>`. Maybe I am using some old C++ rules, but does the callee resolution differentiate result types?

Reply:
If there were naming conflicts, they probably would have been present before this change. The callee does not differentiate result types though. Either way, I changed it to `getFirstLoweringConfig` to be clearer about the functionality and also avoid any such issues.